python/pyarrow/parquet.py

```diff
@@ -104,6 +111,360 @@ def _get_column_indices(self, column_names):
                     for column in column_names]
 
 
+# ----------------------------------------------------------------------
+# Metadata container providing instructions about reading a single Parquet
+# file, possibly part of a partitioned dataset
+
+
+class ParquetDatasetPiece(object):
+    """
+    A single chunk of a potentially larger Parquet dataset to read. The
+    arguments will indicate to read either a single row group or all row
+    groups, and whether to add partition keys to the resulting pyarrow.Table
+
+    Parameters
+    ----------
+    partition_keys : list of tuples
+      [(column name, ordinal index)]
+    """
+
+    def __init__(self, path, row_group=None, partition_keys=None):
+        self.path = path
+        self.row_group = row_group
+        self.partition_keys = partition_keys or []
+
+    def __eq__(self, other):
+        if not isinstance(other, ParquetDatasetPiece):
+            return False
+        return (self.path == other.path and
+                self.row_group == other.row_group and
+                self.partition_keys == other.partition_keys)
+
+    def __ne__(self, other):
+        return not (self == other)
+
+    def __repr__(self):
+        return ('{0}({1!r}, row_group={2!r}, partition_keys={3!r})'
+                .format(type(self).__name__, self.path,
+                        self.row_group,
+                        self.partition_keys))
+
+    def __str__(self):
+        result = ''
+
+        if len(self.partition_keys) > 0:
+            partition_str = ', '.join('{0}={1}'.format(name, index)
+                                      for name, index in self.partition_keys)
+            result += 'partition[{0}] '.format(partition_str)
+
+        result += self.path
+
+        if self.row_group is not None:
+            result += ' | row_group={0}'.format(self.row_group)
+
+        return result
+
+    def open(self, open_file_func):
+        return open_file_func(self.path)
+
+    def read(self, open_file_func, columns=None, nthreads=1,
+             partitions=None):
+        reader = self.open(open_file_func)
```

Comment:
maybe ``open_file_func=None`` convenient as a default here? (maybe also change the signature
``columns=None, partitions=None, nthreads=1, open_file_func=None``?

indifferent tweeen ``partitions`` and ``columns`` ordering in signature, but should be consistent (e.g. ``read_row_group``)